syntax = "proto3";

service EngineServer {
    rpc Join(JoinRequest) returns (JoinResponse) {}

    rpc Leave(LeaveRequest) returns (LeaveResponse) {}

    rpc GetPlayers(GetPlayersRequest) returns (GetPlayersResponse) {}

    rpc Start(StartRequest) returns (StartResponse) {}

    rpc Kill(KillRequest) returns (KillResponse) {}

    rpc Check(CheckRequest) returns (CheckResponse) {}

    rpc Vote(VoteRequest) returns (VoteResponse) {}

    rpc EndDay(EndDayRequest) returns (EndDayResponse) {}

    rpc EndNight(EndNightRequest) returns (EndNightResponse) {}

    rpc PublishSheriffChecks(PublishRequest) returns (PublishResponse) {}

    rpc GameInfo(InfoRequest) returns (stream InfoResponse) {}

    rpc Chat(ChatRequest) returns (ChatResponse) {}
}

message JoinRequest {
    string name = 1;
    optional string text = 2;
}

message JoinResponse {
    int32 id = 1;
    optional string text = 2;
}

message LeaveRequest {
    string name = 1;
    optional string text = 2;
}

message LeaveResponse {
    optional string text = 2;
}

message GetPlayersRequest {
    string name = 1;
    optional string text = 2;
}

message GetPlayersResponse {
    string text = 1;
    string names = 2;
}

message StartRequest {
    string name = 1;
    optional string text = 2;
}

message StartResponse {
    bool started = 1;
    string role = 2;
    optional string text = 3;
    optional string players = 4;
    optional string mafias = 5;
}

message InfoRequest {
    string name = 1;
    optional string text = 2;
}

message InfoResponse {
    string type = 1;
    optional string text = 2;
    optional string dead_player_name = 3;
}

message KillRequest {
    string name = 1;
    string kill_name = 2;
    optional string text = 3;
}

message KillResponse {
    bool result = 1;
    optional string text = 2;
}

message CheckRequest {
    string name = 1;
    string check_name = 2;
    optional string text = 3;
}

message CheckResponse {
    bool result = 1;
    optional string role = 2;
    optional string text = 3;
}

message VoteRequest {
    string name = 1;
    string vote_name = 2;
    optional string text = 3;
}

message VoteResponse {
    bool result = 1;
    optional string text = 2;
}

message EndDayRequest {
    string name = 1;
    optional string text = 2;
}

message EndDayResponse {
    bool ended = 1;
    optional string text = 2;
    optional string dead_player_name = 3;
}

message EndNightRequest {
    string name = 1;
    optional string text = 2;
}

message EndNightResponse {
    bool ended = 1;
    optional string text = 2;
    optional string dead_player_name = 3;
}

message PublishRequest {
    string name = 1;
    optional string text = 2;
}

message PublishResponse {
    bool result = 1;
    optional string text = 2;
}

message ChatRequest {
    string name = 1;
    string message = 2;
}
 
message ChatResponse {
    optional bool result = 1;
    optional string text = 2;
}